### PR TITLE
test/alternator: functional tests for Alternator multi-item transaction feature

### PR DESCRIPTION
**Pull request by nyh**
This patch adds extensive functional tests for the DynamoDB multi-item transactions feature - the TransactWriteItems and TransactGetItems requests. We add 43 test functions, spanning more than 1000 lines of code, covering the different parameters and corner cases of these requests.
    
Because we don't support the transaction feature in Alternator yet (this is issue #5064), all of these tests fail on Alternator but all of them were tested to pass on DynamoDB. So all new tests are marked "xfail".
    
These tests will be handy for whoever will implement this feature as an acceptance test, and can also be useful for whoever will just want to understand this feature better - the tests are short and simple and heavily commented. 
    
Note that these tests only check the correct functionality of individual calls of these requests - these tests cannot and do not check the consistency or isolation guarantees of concurrent invocations of several requests. Such tests would require a different test framework, such as the one requested in issue #6350, and are therefore not part of this patch.
    
Note that this patch includes ONLY tests, and does not mean that an implementation of the feature will soon follow. In fact, nobody is currently working on implementing this feature.

### CODE REVIEW

**Comment by nyh** [comment]
Copyedited and pushed a new version.
The newest version just has (a lot...) of typo fixes in comments, no functional changes to the tests.

**Comment by nyh** [comment]
Please review and consider merging. This is a "harmless" set of xfailing tests (passing on DynamoDB, failing on Alternator) for the "transactions" feature that we don't have yet. I think that it can really help anyone who wants to understand this feature better (as it explores the various corner cases of the DynamoDB API) and certainly for anyone in the future who might want to implement it.

File: test/alternator/test_transact.py [hunk 1]
```diff
@@ -0,0 +1,974 @@
...
+        ret = test_table_s.meta.client.transact_get_items(TransactItems=[
+            { 'Get': {
+                'TableName': test_table_s.name,
+                'Key': {'p': p},
+                'ProjectionExpression': '#xx,#zz',
+                # The reference "#zz" is missing in ExpressionAttributeNames
+                'ExpressionAttributeNames': {'#xx': 'x'}
+            }}])
+
+# Test the ability to read multiple items in one transaction. We can actually
```
Comment (swasik) [inline test/alternator/test_transact.py:940]: You have the same for the write somewhere in the middle of the file. Maybe we should keep it together for easier reading.

File: test/alternator/test_transact.py [hunk 2]
```diff
@@ -0,0 +1,974 @@
...
+                'TableName': test_table_s.name,
+                'Key': {'p': p2},
+                'ConditionExpression': 'attribute_not_exists(p)'
+            }}])
+
+# Test that a transaction may involve more than one table, not just more than
+# one item.
+@pytest.mark.xfail(reason="#5064 - transactions not yet supported")
+def test_transact_write_items_multi_table_true(test_table_s, test_table):
+    p1 = random_string()
```
[the remark this answers is not in the document]
Comment (nyh) [reply test/alternator/test_transact.py:435]: Oh, the "_s" just means the key is a single string. I also have "_sn" (the key is a string and a number) and a bunch of others, defined in conftest.py. The idea is that if we have a handful of tables shared by many tests, each individual test doesn't even need to create a table - which is the slowest part of each test. Thanks to this approach, many of these tests take milliseconds (!) to run.

Anyway, "test_table", without the prefix, was just the very first table I created in conftest.py. Actually, I do have another name, `test_table_ss` defined in conftest.py. Maybe I should indeed use that name for clarity, not test_table.

(by the way, it was also a mistake to name these fixtures test_something - it makes them sound like they are tests! I invented this name before I became a pytest expert... But it's not that terrible so I never went back to rename them in 1000 tests).

File: test/alternator/test_transact.py [hunk 3]
```diff
@@ -0,0 +1,974 @@
...
+@pytest.mark.xfail(reason="#5064 - transactions not yet supported")
+def test_transact_write_items_100(test_table_s):
+    p = random_string()
+    items = [{'p': p + str(i), 'x': i} for i in range(100)]
+    test_table_s.meta.client.transact_write_items(TransactItems=[
+        { 'Put': {
+            'TableName': test_table_s.name,
+            'Item': item,
+        }} for item in items])
+    # verify that all 100 items went it
```
[the remark this answers is not in the document]
Comment (nyh) [reply test/alternator/test_transact.py:476]: Right.

File: test/alternator/test_transact.py [hunk 4]
```diff
@@ -0,0 +1,974 @@
...
+    assert 'CancellationReasons' in e.value.response
+    reasons = e.value.response['CancellationReasons']
+    # The CancellationReasons are listed in the same order of the actions
+    # in the transaction - we expect the first two actions to have not
+    # failed, so they have Code="None", and the third and forth both failed.
+    assert len(reasons) == 4
+    assert reasons[0]['Code'] == 'None'
+    assert reasons[1]['Code'] == 'None'
+    assert reasons[2]['Code'] == 'ConditionalCheckFailed'
+    assert reasons[3]['Code'] == 'ConditionalCheckFailed'
```
[the remark this answers is not in the document]
Comment (nyh) [reply test/alternator/test_transact.py:575]: Yes, all the tests passed on DynamoDB (in fact, that's the only thing I could run them on, since we don't have this feature on Alternator). 

And yes, it seems that all conditions are checked before making a decision, there is no "short circuiting" like in C's "&&". However, as I uncovered in other tests, if the condition has a **mistake** it's not just false (syntax error, wrong type, etc.) it is possible that only the first error is recognized, and in some cases we don't even get a list of errors and just get a single failure on the entire transaction. I'm not sure it will be very important to replicate 100% each and every one of these details in our own implementation, but in general it's a good idea to understand the rules before breaking them. If after careful consideration we decide that in our implementation we will deviate from one DynamoDB's behaviors, I can change the test to accept both behaviors, or remove the test, and so on. When I (rarely) do that, I usually add a long comment in the test, so the test suite becomes an interesting read for people wondering how DynamoDB, Alternator, or both, behave - and why.

File: test/alternator/test_transact.py [hunk 5]
```diff
@@ -0,0 +1,974 @@
...
+    with pytest.raises(ClientError, match='TransactionCanceledException') as e:
+        test_table_s.meta.client.transact_write_items(TransactItems=[
+            { 'Put': {
+                'TableName': test_table_s.name,
+                # The key p should be a string, not a number, so this
+                # item has a validation error.
+                'Item': {'p': 7}
+            }}])
+    reasons = e.value.response['CancellationReasons']
+    assert reasons[0]['Code'] == 'ValidationError'
```
[the remark this answers is not in the document]
Comment (nyh) [reply test/alternator/test_transact.py:663]: I was lazy to check - I didn't think it was interesting because I checked the message in other tests that covered more than this test does, and the message wasn't it's focus.
But I can add the message check here too. I'll do it.

File: test/alternator/test_transact.py [hunk 6]
```diff
@@ -0,0 +1,974 @@
...
+            { 'Put': {
+                'TableName': test_table_s.name,
+                'Item': {'p': p},
+                'ConditionExpression': 'attribute_exists(p)',
+                # ExpressionAttributeNames expects strings (names of
+                # attributes), not the integer 7.
+                'ExpressionAttributeNames': {'#xyz': 7}
+            }}])
+
+# TODO: check size limits on total write size (we already checked above
```
[the remark this answers is not in the document]
Comment (nyh) [reply test/alternator/test_transact.py:835]: Not really. It's slower and more expensive than other tests because if X is the limit, we need to also check that a request with X-1 bytes does work. But this shouldn't be prohibitive (it will still cost less than a cent to run).
I'll do it, just to get rid of the last TODO :-) (as you can see, this test file has 1000 lines of code - it started with dozens of TODO lines and then I wrote the tests one by one until this was just the last one left).

File: test/alternator/test_transact.py (hunk 1 above)
Comment (nyh) [reply test/alternator/test_transact.py:940]: Yes, but on the other hand I decided to bunch all the read tests together and the all the write tests together, so this is why this test is together with the other write tests. So I'm not sure which of the two classifications is more helpful. Maybe you're right, I'll think about it again.

File: test/alternator/test_transact.py [hunk 7]
```diff
@@ -0,0 +1,974 @@
...
+                'Key': {'p': p},
+                # The condition attribute_exists(p) is false for an item that
+                # doesn't exist
+                'ConditionExpression': 'attribute_exists(p)'
+            }}])
+
+# Single Update action without a condition
+# Unlike the previous test_transact_write_items_single_* tests which used
+# trivial conditions without ExpressionAttributeNames and
+# ExpressionAttributeValues, the following tests do use those (this test
```
[the remark this answers is not in the document]
Comment (nyh) [reply test/alternator/test_transact.py:174]: I'll do that.
Note that this point (the last three lines of this comment) isn't really an important "section" of the test. It's just a comment that although in the previous tests we simplified the API usage by not using ExpressionAttributeNames/Values, we also need to start testing those so we have them in the following tests.

File: test/alternator/test_transact.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (nyh) [reply test/alternator/test_transact.py:435]: I changed the usage of "test_table" in this test to "test_table_ss", which I agree is more self-descriptive.
There are 142 existing tests using the name "test_table", I won't be changing all of them in this patch (obviously).  We can consider doing this later, I'll open an issue about it.

[the remark this answers is not in the document]
Comment (nyh) [reply test/alternator/test_transact.py:435]: https://github.com/scylladb/scylladb/issues/22494

File: test/alternator/test_transact.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (nyh) [reply test/alternator/test_transact.py:575]: I'll add some comments about this to the "book" in the test ;-)

File: test/alternator/test_transact.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (nyh) [reply test/alternator/test_transact.py:663]: Added the check.

File: test/alternator/test_transact.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (nyh) [reply test/alternator/test_transact.py:835]: By the way if you're curious about the "more expensive" part, well, I calculated writing 3MB (to verify that 3MB is allowed) in a transaction costs 6,000 WCU (that's 2 WCU per 1 KB), which at $0.625 per million WCU (the new lowered prices) costs 0.375 cents. 
I think we'll survive that cost :-) Especially that I'll probably only ever run this test 10 or so times.

**Comment by nyh** [comment]
Pushed a new version addressing all review comments (see my individual responses), and adding 4 more tests - 3 covering the transaction size limit that previously was a TODO (now there are no more TODOs), and a fourth testing that a Put action replaces an item (that's "obvious" seeing that's how PutItem behaves, but in Scylla it is far from obvious because that's not how a CQL "Insert" behaves).

**Review by swasik — APPROVED** [review]

**Comment by nyh** [comment]
@scylladb/scylla-maint please consider merging.
